src/markov/markov_approx.jl

```diff
@@ -214,3 +218,402 @@ function estimate_mc_discrete{T}(X::Vector{T})
 
     return estimate_mc_discrete(X, states)
 end
+
+"""
+Compute a finite-state Markov chain approximation to a VAR(1) process of the form
+    ```math
+        y_{t+1} = b + By_{t} + \Psi^(\frac{1}{2})*\epsilon_{t+1}
+    ```
+where `\epsilon_{t+1}` is an (M x 1) vector of independent standard normal
+innovations
+    ```julia
+    P, X = discrete_var(b, B, Psi, Nm, nMoments, method, nSigmas)
+    ```
+##### Arguments
+- `b::ScalarOrArray` : (M x 1) constant vector
+                       (M=1 corresponds scalar case)
+- `B::ScalarOrArray` : (M x M) matrix of impact coefficients
+- `Psi::ScalarOrArray` : (M x M) variance-covariance matrix of the innovations
+- `Nm::Integer` : Desired number of discrete points in each dimension
+##### Optional
+- `nMoments::Integer` : Desired number of moments to match. The default is 2.
+- `method::Symbol` : Symbol specifying the method used to determine the grid
+             points. Accepted inputs are `:even`. Please see the
+             paper for more details.
+             NOTE: `:quantile` and `:quadrature` are not supported now.
+- `nSigmas::Union{Void, TI}` : If the `:even` option is specified, nSigmas is used to
+              determine the number of unconditional standard deviations
+              used to set the endpoints of the grid. The default is
+              sqrt(Nm-1).
+
+##### Returns
+- `P` : (Nm^M x Nm^M) probability transition matrix. Each row
+        corresponds to a discrete conditional probability
+        distribution over the state M-tuples in X
+- `X` : (M x Nm^M) matrix of states. Each column corresponds to an
+        M-tuple of values which correspond to the state associated
+        with each row of P
+
+##### NOTES
+- discrete_var only accepts non-singular variance-covariance matrices.
+- discrete_var only constructs tensor product grids where each dimension
+  contains the same number of points. For this reason it is recommended
+  that this code not be used for problems of more than about 4 or 5
+  dimensions due to curse of dimensionality issues.
+- Future updates will allow for singular variance-covariance matrices and
+  sparse grid specifications.
+"""
+function discrete_var{TI<:Integer}(b::ScalarOrArray, B::ScalarOrArray,
+                     Psi::ScalarOrArray, Nm::TI,
+                     nMoments::TI=2, method::Symbol=:even,
+                     nSigmas::Union{Void, TI}=nothing)
+
+    if typeof(B) <: Real
```

Comment:
@cc7768 I tried to do that before but it was not straightforward for some reason (which I forgot). Let me carefully work on the stuff regarding dispatching next week.

Another question I have is efficiency. The benefit of following your strategy that scalar `b` is passed to a general function as `b=ones(M)*b` is we can avoid `if...else..` style branching. However, this makes all calculation `1x1` matrix operations instead of scalar ones. I am not sure but it might be faster to deal with scalar as scalar, because this function requires many matrix operations for VAR and `1x1` matrix operations are more time-consuming than scalar operations. Do you have any thoughts?